### PR TITLE
Cart Checkout

### LINKED ISSUE #57: Multiple Doughnut Orders
Modify `Order` so that an Order with multiple donuts/quantities can be created

### LINKED ISSUE #58: Cart Checkout
Create Carts checkout feature

### PR DESCRIPTION
**Pull request by chrisyang365**
`GET /api/user/:id/orders`
- gets the order history for user with `:id`
- response format will look like the following: 
```
[
    {
        "id": 1,
        "created_at": "2021-11-14T04:24:14.914Z",
        "updated_at": "2021-11-14T04:24:14.914Z",
        "user": {
            "username": "user1"
        },
        "order_items": [
            {
                "quantity": 10,
                "doughnut": {
                    "name": "donut1",
                    "price": 1231.0,
                    "description": "test_dsec",
                    "quantity": 1231
                }
            },
            {
                "quantity": 10,
                "doughnut": {
                    "name": "donut2",
                    "price": 3451231.0,
                    "description": "test_dsec2",
                    "quantity": 211231
                }
            }
        ]
    },
    {
        "id": 2,
        "created_at": "2021-11-14T04:33:08.154Z",
        "updated_at": "2021-11-14T04:33:08.154Z",
        "user": {
            "username": "user1"
        },
        "order_items": [
            {
                "quantity": 20,
                "doughnut": {
                    "name": "donut1",
                    "price": 1231.0,
                    "description": "test_dsec",
                    "quantity": 1231
                }
            },
            {
                "quantity": 30,
                "doughnut": {
                    "name": "donut2",
                    "price": 3451231.0,
                    "description": "test_dsec2",
                    "quantity": 211231
                }
            }
        ]
    },
    {
        "id": 3,
        "created_at": "2021-11-15T01:32:42.348Z",
        "updated_at": "2021-11-15T01:32:42.348Z",
        "user": {
            "username": "user1"
        },
        "order_items": [
            {
                "quantity": 69,
                "doughnut": {
                    "name": "donut2",
                    "price": 3451231.0,
                    "description": "test_dsec2",
                    "quantity": 211231
                }
            }
        ]
    },
    {
        "id": 4,
        "created_at": "2021-11-15T02:39:11.016Z",
        "updated_at": "2021-11-15T02:39:11.016Z",
        "user": {
            "username": "user1"
        },
        "order_items": [
            {
                "quantity": 1010,
                "doughnut": {
                    "name": "donut2",
                    "price": 3451231.0,
                    "description": "test_dsec2",
                    "quantity": 211231
                }
            }
        ]
    },
    {
        "id": 5,
        "created_at": "2021-11-15T02:39:38.233Z",
        "updated_at": "2021-11-15T02:39:38.233Z",
        "user": {
            "username": "user1"
        },
        "order_items": [
            {
                "quantity": 1010,
                "doughnut": {
                    "name": "donut2",
                    "price": 3451231.0,
                    "description": "test_dsec2",
                    "quantity": 211231
                }
            }
        ]
    },
    {
        "id": 6,
        "created_at": "2021-11-15T02:39:56.877Z",
        "updated_at": "2021-11-15T02:39:56.877Z",
        "user": {
            "username": "user1"
        },
        "order_items": [
            {
                "quantity": 1111,
                "doughnut": {
                    "name": "donut2",
                    "price": 3451231.0,
                    "description": "test_dsec2",
                    "quantity": 211231
                }
            }
        ]
    },
    {
        "id": 7,
        "created_at": "2021-11-15T02:45:09.577Z",
        "updated_at": "2021-11-15T02:45:09.577Z",
        "user": {
            "username": "user1"
        },
        "order_items": [
            {
                "quantity": 15049,
                "doughnut": {
                    "name": "donut2",
                    "price": 3451231.0,
                    "description": "test_dsec2",
                    "quantity": 211231
                }
            },
            {
                "quantity": 6969,
                "doughnut": {
                    "name": "donut1",
                    "price": 1231.0,
                    "description": "test_dsec",
                    "quantity": 1231
                }
            }
        ]
    }
]
```

`POST /api/user/:id/orders`
- creates an Order entity
- requires `doughnut_id` and `quantity` to be specified in the request body like so:
```
{
    "doughnut_id" : 2,
    "quantity" : 10
}
```
- response format will be the same response format as `GET /api/user/:id/orders` except the response will only display the newly-created `Order` entity

`POST /api/user/:id/cart/checkout`
- empties the cart for user with `:id`
- `CartItem` entities are converted to `OrderItem` entities and then deleted to empty the cart
- response format changed to:
```
[
    {
        "id": 4,
        "created_at": "2021-11-15T02:45:17.056Z",
        "updated_at": "2021-11-15T02:45:17.056Z",
        "user": {
            "username": "user1"
        },
        "cart_items": [
            {
                "quantity": 6969,
                "doughnut": {
                    "name": "donut1",
                    "price": 1231.0,
                    "description": "test_dsec",
                    "quantity": 1231
                }
            },
            {
                "quantity": 1212,
                "doughnut": {
                    "name": "donut2",
                    "price": 3451231.0,
                    "description": "test_dsec2",
                    "quantity": 211231
                }
            }
        ]
    }
]
```

### CODE REVIEW

File: app/controllers/api/cart_controller.rb [hunk 1]
```diff
@@ -38,6 +53,22 @@ def destroy
     end
   end
 
+  def checkout
+    order = Order.create(user: @user)
+    cart_items = CartItem.where(cart_id: @cart.id)
+    if !cart_items.empty?
+      cart_items_json = cart_items.as_json(only: [:quantity, :doughnut_id])
+      puts "this is cart_items_json"
+      puts cart_items_json
```
Comment (aroddick) [inline app/controllers/api/cart_controller.rb]: Could we remove this?

File: app/controllers/api/orders_controller.rb [hunk 2]
```diff
@@ -1,31 +1,47 @@
...
+    orders = Order.includes(order_items: [:doughnut]).where(user: @user)
+    render json: orders, except: [:user_id],
+      include: [:user => {:only => [:username]}, :order_items => {:only => [:quantity, :doughnut], :include => [:doughnut => {:only => [:name, :price, :description, :quantity]}]}]
   end
 
   def create
-    order = Order.create(order_params)
-    if order.id
-      render json: order, include: [:user => {:only => [:id, :username]}, :doughnut => {:only => [:id, :name, :price, :description]}], only: [:id, :created_at, :updated_at, :user, :doughnut]
+    order = Order.create(user: @user)
```
Comment (aroddick) [inline app/controllers/api/orders_controller.rb:17]: This seems to create an order entity even if there is an error with saving the order_item. Is that what you want?

File: app/controllers/api/cart_controller.rb [hunk 3]
```diff
@@ -38,6 +53,22 @@ def destroy
...
+      cart_items_json.each {|item| item["order_id"] = order.id}
+      OrderItem.create(cart_items_json)
+      @cart.delete
```
Comment (aroddick) [inline app/controllers/api/cart_controller.rb:63]: I am getting this error when attempting to checkout cart:

> web_1  | ActiveRecord::InvalidForeignKey (PG::ForeignKeyViolation: ERROR:  update or delete on table "carts" violates foreign key constraint "fk_rails_6cdb1f0139" on table "cart_items"
web_1  | DETAIL:  Key (id)=(35) is still referenced from table "cart_items".
web_1  | ):
web_1  |   
web_1  | app/controllers/api/cart_controller.rb:65:in `checkout'

Do we need to delete the cart_items first?

File: app/controllers/api/cart_controller.rb [hunk 4]
```diff
@@ -38,6 +53,22 @@ def destroy
     end
   end
 
+  def checkout
+    order = Order.create(user: @user)
```
Comment (aroddick) [inline app/controllers/api/cart_controller.rb]: This also seems to create an Order entity even if the cart is empty.

File: app/controllers/api/cart_controller.rb [hunk 5]
```diff
@@ -2,25 +2,40 @@ class Api::CartController < ApplicationController
...
-    render json: items
+    ### COMPARE FOR LOAD TESTING ###
+    # items = CartItem
+    #   .select('cart_items.*, doughnuts.name, doughnuts.price, doughnuts.description')
+    #   .joins(:cart, :doughnut)
+    #   .where(cart_id: @cart.id)
+    # render json: items
+    @cart = Cart.includes(cart_items: [:doughnut]).where(user_id: @user.id)
+    render json: @cart, except: [:user_id],
+      include: [:user => {:only => [:username]}, :cart_items => {:only => [:quantity, :doughnut], :include => [:doughnut => {:only => [:name, :price, :description, :quantity]}]}]
```
Comment (aroddick) [inline app/controllers/api/cart_controller.rb]: Could we also include the cart_item ID so the client can delete the doughnut from cart?

File: app/controllers/api/cart_controller.rb (hunk 1 above)
Comment (chrisyang365) [reply app/controllers/api/cart_controller.rb]: oops yeah

**Pull request merged**